File: 0061-rotate-list/0061-rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if head is None:
            return head
        curr, length, i = head, 0, 0
        while curr:
            length += 1
            curr = curr.next
        headIdx = (length - (k%length)) % length
        if headIdx == 0:
            return head
        curr = head
        while i < headIdx-1:
            curr = curr.next
            i += 1
        last, newHead = curr, curr.next
        last.next = None
        curr = newHead
        while curr.next:
            curr = curr.next
        curr.next = head
        return newHead
        

File: 0061-rotate-list/0061-rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *   int val;
 *   ListNode next;
 *   ListNode() {}
 *   ListNode(int val) { this.val = val; }
 *   ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
  public ListNode rotateRight(ListNode head, int k) {
    if (head == null) return head;
    ListNode curr = head;
    int len = 0;
    int i = 0;
    while (curr != null) {
      curr = curr.next;
      len++;
    }
    int headIdx = (len - (k % len)) % len;
    if (headIdx == 0) return head;
    curr = head;
    while (i < headIdx - 1) {
      curr = curr.next;
      i++;
    }
    ListNode last = curr;
    ListNode newHead = curr.next;
    last.next = null;
    curr = newHead;
    while (curr.next != null) curr = curr.next;
    curr.next = head;
    return newHead;
  }
}